frontend/packages/kubevirt-plugin/src/components/modals/add-template-source/add-template-source.tsx

```diff
@@ -0,0 +1,518 @@
+import * as React from 'react';
+import {
+  provisionerAccessModeMapping,
+  getAccessModeForProvisioner,
+  dropdownUnits,
+  accessModeRadios,
+} from '@console/internal/components/storage/shared';
+import {
+  StorageClassResourceKind,
+  k8sCreate,
+  TemplateKind,
+  PersistentVolumeClaimKind,
+} from '@console/internal/module/k8s';
+import {
+  ModalComponentProps,
+  ModalTitle,
+  ModalBody,
+  createModalLauncher,
+  ModalFooter,
+} from '@console/internal/components/factory';
+import { useK8sWatchResource } from '@console/internal/components/utils/k8s-watch-hook';
+import { StorageClassModel, PersistentVolumeClaimModel } from '@console/internal/models';
+import {
+  Form,
+  FileUpload,
+  ExpandableSection,
+  FormSelectOption,
+  Alert,
+  Button,
+  Select,
+  SelectOption,
+  TextInput,
+  FormSelect,
+  ActionGroup,
+  Checkbox,
+} from '@patternfly/react-core';
+import {
+  RequestSizeInput,
+  LoadingInline,
+  useAccessReview2,
+  LoadingBox,
+} from '@console/internal/components/utils';
+import { UploadPVCFormStatus } from '../../cdi-upload-provider/upload-pvc-form/upload-pvc-form-status';
+import { getDefaultStorageClass } from '../../../selectors/config-map/sc-defaults';
+import { createUploadPVC } from '../../../k8s/requests/cdi-upload/cdi-upload-requests';
+import { DataVolumeWrapper } from '../../../k8s/wrapper/vm/data-volume-wrapper';
+import {
+  DataVolumeSourceType,
+  AccessMode,
+  TEMPLATE_BASE_IMAGE_NAME_PARAMETER,
+  TEMPLATE_BASE_IMAGE_NAMESPACE_PARAMETER,
+} from '../../../constants';
+import { EXAMPLE_CONTAINER, FEDORA_IMAGE_LINK, RHEL_IMAGE_LINK } from '../../../utils/strings';
+import { FormRow } from '../../form/form-row';
+import ProjectDropdown from '../../form/ProjectDropdown';
+import { getParameterValue } from '../../../selectors/selectors';
+import { V1alpha1DataVolume } from '../../../types/vm/disk/V1alpha1DataVolume';
+import { DataVolumeModel } from '../../../models';
+import { CDIUploadContextProps } from '../../cdi-upload-provider/cdi-upload-provider';
+
+import './add-template-source.scss';
+import { PVCSelector } from '../../create-vm/pvc-selector';
+
+const DATA_SOURCE = {
+  UPLOAD: 'Upload local file (creates PVC)',
+  URL: 'Import via URL (creates PVC)',
+  REGISTRY: 'Import via Registry (creates PVC)',
+  CLONE: 'Clone existing PVC',
+};
+
+const DATA_SOURCES = [
```

Comment:
we should probably create a new `SelectDropdownObjectEnum` for this object.
see `storage-ui-source.ts` for example.

Reply:
done!